README.md
XenIface - The Xen Interface Driver for Windows
===============================================

The XenIface package consists of a single device driver:

*    xeniface.sys is a driver which attaches to a virtual device created
     by XenBus and provides a WMI to xenstore (and also an IOCTL interface
     for simple xenstore read/write access).

Quick Start Guide
=================

Building the driver
-------------------

First you'll need a device driver build environment for Windows 8. For this
you must use:

*   Visual Studio 2012 (Professional or Ultimate)
*   Windows Driver Kit 8

(See http://msdn.microsoft.com/en-us/windows/hardware/hh852365.aspx). You
may find it useful to install VirtualCloneDrive from http://www.slysoft.com
as Visual Studio is generally supplied in ISO form.

Install Visual Studio first (you only need install MFC for C++) and then
the WDK. Set an environment variable called VS to the base of the Visual
Studio Installation (e.g. C:\Program Files\Microsoft Visual Studio 11.0) and
a variable called KIT to the base of the WDK
(e.g. C:\Program Files\Windows Kits\8.0). Also set an environment variable
called SYMBOL\_SERVER to point at a location where driver symbols can be
stored. This can be local directory e.g. C:\Symbols.

Next you'll need a 3.x version of python (which you can get from
http://www.python.org). Make sure python.exe is somewhere on your default
path.

Now fire up a Command Prompt and navigate to the base of your git repository.
At the prompt type:

    build.py checked

This will create a debug build of the driver. To create a non-debug build
type:

    build.py free

Installing the driver
---------------------

See INSTALL.md

API Versions
============

It is important that introduction of a new API, introduction of a new
version of an existing API or retirement of an old version of an API is
managed carefully to avoid incompatibilities between clients and
providers. The general API versioning policy is described below:

Each distinct set of API versions exported by a driver maps to a PDO
revision. The DeviceID of each PDO created by the driver will specify the
latest revision supported and all others will be contained within the
HardwareIDs and CompatibleIDs.
Hence, when a new version of an API is added, a new PDO revision will be
added. When a version of an API is removed then ALL revisions that API
version maps to will be removed.

To avoid a situation where a new version of the package is installed that
is incompatible with any child drivers that make use of the APIs, each
child 'subscribes' to an API by writing a registry value with the version
number of that API that they consume into a registry key under the service
key of the providing driver. E.g. if driver 'foo' consumes version 1 of
driver 'bar''s 'widget' API, then it will write
HLKM/CurrentControlSet/Services/BAR/Interfaces/FOO/WIDGET with the value 1.
The 'bar' package co-installer can then check, prior to installation of a
new version of a driver, that it can still support all of its subscribers.
If any of the API versions subscribed to has been removed then installation
will be vetoed until all the subscriber drivers have been updated to use
the newer versions of the APIs exported by the newer providing driver.

Miscellaneous
=============

For convenience the source repository includes some other scripts:

kdfiles.py
----------

This generates two files called kdfiles32.txt and kdfiles64.txt which can
be used as map files for the .kdfiles WinDBG command.

sdv.py
------

This runs Static Driver Verifier on the source.

clean.py
--------

This removes any files not checked into the repository and not covered by
the .gitignore file.


CHANGELOG
7.0.0 (2013-05-30):
* Initial public release

7.1.0 (2013-07-08):
* New 'bottom layer' of code which can be open sourced.
* Upped version number to reflect build number reset

7.2.0 (2013-09-27):
* Drivers compatible with upstream Xen, QEMU and Linux dom0

8.0.0 (2014-08-15):
* Drivers with new API version scheme

8.1.0 (2015-07-24):
* Revised API versioning and unplug mechanism


build.py
#!python -u

import os, sys
import datetime
import re
import glob
import tarfile
import subprocess
import shutil
import time

def next_build_number():
    try:
        file = open('.build_number', 'r')
        build_number = file.read()
        file.close()
    except IOError:
        build_number = '0'

    file = open('.build_number', 'w')
    file.write(str(int(build_number) + 1))
    file.close()

    return build_number


def make_header():
    now = datetime.datetime.now()

    file = open('include\\version.h', 'w')

    file.write('#define COMPANY_NAME_STR\t"' + os.environ['COMPANY_NAME'] + '"\n')
    file.write('#define PRODUCT_NAME_STR\t"' + os.environ['PRODUCT_NAME'] + '"\n')
    file.write('\n')

    file.write('#define MAJOR_VERSION\t' + os.environ['MAJOR_VERSION'] + '\n')
    file.write('#define MAJOR_VERSION_STR\t"' + os.environ['MAJOR_VERSION'] + '"\n')
    file.write('\n')

    file.write('#define MINOR_VERSION\t' + os.environ['MINOR_VERSION'] + '\n')
    file.write('#define MINOR_VERSION_STR\t"' + os.environ['MINOR_VERSION'] + '"\n')
    file.write('\n')

    file.write('#define MICRO_VERSION\t' + os.environ['MICRO_VERSION'] + '\n')
    file.write('#define MICRO_VERSION_STR\t"' + os.environ['MICRO_VERSION'] + '"\n')
    file.write('\n')

    file.write('#define BUILD_NUMBER\t' + os.environ['BUILD_NUMBER'] + '\n')
    file.write('#define BUILD_NUMBER_STR\t"' + os.environ['BUILD_NUMBER'] + '"\n')
    file.write('\n')

    file.write('#define YEAR\t' + str(now.year) + '\n')
    file.write('#define YEAR_STR\t"' + str(now.year) + '"\n')
    file.write('\n')

    file.write('#define MONTH\t' + str(now.month) + '\n')
    file.write('#define MONTH_STR\t"' + str(now.month) + '"\n')
    file.write('\n')

    file.write('#define DAY\t' + str(now.day) + '\n')
    file.write('#define DAY_STR\t"' + str(now.day) + '"\n')
    file.write('\n')

    file.write('#define OBJECT_PREFIX_STR\t"' + os.environ['OBJECT_PREFIX'] + '"\n')
    file.write('#define OBJECT_GUID(_Name)\t' + os.environ['OBJECT_PREFIX'] + ' ## _Name ## _GUID\n')

    file.close()


def copy_inf(name, vs):
    src = open('src\\%s.inf' % name, 'r')
    dst = open(vs+'\\%s.inf' % name, 'w')

    for line in src:
        line = re.sub('@MAJOR_VERSION@', os.environ['MAJOR_VERSION'], line)
        line = re.sub('@MINOR_VERSION@', os.environ['MINOR_VERSION'], line)
        line = re.sub('@MICRO_VERSION@', os.environ['MICRO_VERSION'], line)
        line = re.sub('@BUILD_NUMBER@', os.environ['BUILD_NUMBER'], line)
        line = re.sub('@COMPANY_NAME@', os.environ['COMPANY_NAME'], line)
        line = re.sub('@PRODUCT_NAME@', os.environ['PRODUCT_NAME'], line)
        dst.write(line)

    dst.close()
    src.close()


def copy_mof(name):
    src = open('src\\%s.mof' % name, 'r')
    dst = open('src\\%s\\wmi.mof' % name, 'w')

    for line in src:
        line = re.sub('@OBJECT_PREFIX@', os.environ['OBJECT_PREFIX'], line)
        dst.write(line)

    dst.close()
    src.close()


def get_expired_symbols(name, age = 30):
    path = os.path.join(os.environ['SYMBOL_SERVER'], '000Admin\\history.txt')

    try:
        file = open(path, 'r')
    except IOError:
        return []

    threshold = datetime.datetime.utcnow() - datetime.timedelta(days = age)

    expired = []

    for line in file:
        item = line.split(',')

        if (re.match('add', item[1])):
            id = item[0]
            date = item[3].split('/')
            time = item[4].split(':')
            tag = item[5].strip('"')

            age = datetime.datetime(year = int(date[2]),
                                    month = int(date[0]),
                                    day = int(date[1]),
                                    hour = int(time[0]),
                                    minute = int(time[1]),
                                    second = int(time[2]))
            if (tag == name and age < threshold):
                expired.append(id)

        elif (re.match('del', item[1])):
            id = item[2].rstrip()
            try:
                expired.remove(id)
            except ValueError:
                pass

    file.close()

    return expired


def get_configuration(release, debug):
    configuration = release

    if debug:
        configuration += ' Debug'
    else:
        configuration += ' Release'

    return configuration


def get_target_path(release, arch, debug, vs):
    configuration = get_configuration(release, debug)
    name = ''.join(configuration.split(' '))
    target = { 'x86': os.sep.join([name, 'Win32']), 'x64': os.sep.join([name, 'x64']) }
    target_path = os.sep.join([vs, target[arch]])

    return target_path


def shell(command, dir):
    print(dir)
    print(command)
    sys.stdout.flush()
    
    sub = subprocess.Popen(' '.join(command), cwd=dir,
                           stdout=subprocess.PIPE,
                           stderr=subprocess.STDOUT)

    for line in sub.stdout:
        print(line.decode(sys.getdefaultencoding()).rstrip())

    sub.wait()

    return sub.returncode


class msbuild_failure(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)

def msbuild(platform, configuration, target, file, args, dir):
    os.environ['PLATFORM'] = platform
    os.environ['CONFIGURATION'] = configuration
    os.environ['TARGET'] = target
    os.environ['FILE'] = file
    os.environ['EXTRA'] = args

    bin = os.path.join(os.getcwd(), 'msbuild.bat')

    status = shell([bin], dir)

    if (status != 0):
        raise msbuild_failure(configuration)


def build_sln(name, release, arch, debug, vs):
    configuration = get_configuration(release, debug)

    if arch == 'x86':
        platform = 'Win32'
    elif arch == 'x64':
        platform = 'x64'

    cwd = os.getcwd()

    msbuild(platform, configuration, 'Build', name + '.sln', '', vs)


def remove_timestamps(path):
    try:
        os.unlink(path + '.orig')
    except OSError:
        pass

    os.rename(path, path + '.orig')

    src = open(path + '.orig', 'r')
    dst = open(path, 'w')

    for line in src:
        if line.find('TimeStamp') == -1:
            dst.write(line)

    dst.close()
    src.close()

def sdv_clean(name, vs):
    path = [vs, name, 'sdv']
    print(path)

    shutil.rmtree(os.path.join(*path), True)

    path = [vs, name, 'sdv.temp']
    print(path)

    shutil.rmtree(os.path.join(*path), True)

    path = [vs, name, 'staticdv.job']
    print(path)

    try:
        os.unlink(os.path.join(*path))
    except OSError:
        pass

    path = [vs, name, 'refine.sdv']
    print(path)

    try:
        os.unlink(os.path.join(*path))
    except OSError:
        pass

    path = [vs, name, 'sdv-map.h']
    print(path)

    try:
        os.unlink(os.path.join(*path))
    except OSError:
        pass


def run_sdv(name, dir, vs):
    configuration = get_configuration('Windows 8', False)
    platform = 'x64'

    msbuild(platform, configuration, 'Build', name + '.vcxproj',
            '', os.path.join(vs, name))

    sdv_clean(name, vs)

    msbuild(platform, configuration, 'sdv', name + '.vcxproj',
            '/p:Inputs="/scan"', os.path.join(vs, name))

    path = [vs, name, 'sdv-map.h']
    file = open(os.path.join(*path), 'r')

    for line in file:
        print(line)

    file.close()

    msbuild(platform, configuration, 'sdv', name + '.vcxproj',
            '/p:Inputs="/check:default.sdv"', os.path.join(vs, name))

    path = [vs, name, 'sdv', 'SDV.DVL.xml']
    remove_timestamps(os.path.join(*path))

    msbuild(platform, configuration, 'dvl', name + '.vcxproj',
            '', os.path.join(vs, name))

    path = [vs, name, name + '.DVL.XML']
    shutil.copy(os.path.join(*path), dir)

    path = [vs, name, 'refine.sdv']
    if os.path.isfile(os.path.join(*path)):
        msbuild(platform, configuration, 'sdv', name + '.vcxproj',
                '/p:Inputs=/refine', os.path.join(vs, name))


def symstore_del(name, age):
    symstore_path = [os.environ['KIT'], 'Debuggers']
    if os.environ['PROCESSOR_ARCHITECTURE'] == 'x86':
        symstore_path.append('x86')
    else:
        symstore_path.append('x64')
    symstore_path.append('symstore.exe')

    symstore = os.path.join(*symstore_path)

    for id in get_expired_symbols(name, age):
        command=['"' + symstore + '"']
        command.append('del')
        command.append('/i')
        command.append(str(id))
        command.append('/s')
        command.append(os.environ['SYMBOL_SERVER'])

        shell(command, None)


def symstore_add(name, release, arch, debug, vs):
    target_path = get_target_path(release, arch, debug, vs)

    symstore_path = [os.environ['KIT'], 'Debuggers']
    if os.environ['PROCESSOR_ARCHITECTURE'] == 'x86':
        symstore_path.append('x86')
    else:
        symstore_path.append('x64')
    symstore_path.append('symstore.exe')

    symstore = os.path.join(*symstore_path)

    version = '.'.join([os.environ['MAJOR_VERSION'],
                        os.environ['MINOR_VERSION'],
                        os.environ['MICRO_VERSION'],
                        os.environ['BUILD_NUMBER']])

    command=['"' + symstore + '"']
    command.append('add')
    command.append('/s')
    command.append(os.environ['SYMBOL_SERVER'])
    command.append('/r')
    command.append('/f')
    command.append('*.pdb')
    command.append('/t')
    command.append(name)
    command.append('/v')
    command.append(version)

    shell(command, target_path)


def manifest():
    cmd = ['git', 'ls-tree', '-r', '--name-only', 'HEAD']

    sub = subprocess.Popen(cmd, stdout=subprocess.PIPE)
    output = sub.communicate()[0]
    ret = sub.returncode

    if ret != 0:
        raise(Exception("Error %d in : %s" % (ret, cmd)))

    return output.decode('utf-8')


def archive(filename, files, tgz=False):
    access='w'
    if tgz:
        access='w:gz'
    tar = tarfile.open(filename, access)
    for name in files :
        try:
            tar.add(name)
        except:
            pass
    tar.close()

def getVsVersion():
    vsenv ={} 
    vars = subprocess.check_output([os.environ['VS']+'\\VC\\vcvarsall.bat', 
                                        '&&', 'set'], 
                                    shell=True)
    for var in vars.splitlines():
        k, _, v = map(str.strip, var.strip().decode('utf-8').partition('='))
        if k.startswith('?'):
            continue
        vsenv[k] = v

    if vsenv['VisualStudioVersion'] == '11.0' :
        return 'vs2012'
    elif vsenv['VisualStudioVersion'] == '12.0' :
        return 'vs2013'

if __name__ == '__main__':
    debug = { 'checked': True, 'free': False }
    sdv = { 'nosdv': False, None: True }
    driver = 'xeniface'
    vs = getVsVersion()

    os.utime('include/version.hx', None)
    os.utime('src/%s.inf' % driver, None)

    if 'COMPANY_NAME' not in os.environ.keys():
        os.environ['COMPANY_NAME'] = 'Xen Project'

    if 'PRODUCT_NAME' not in os.environ.keys():
        os.environ['PRODUCT_NAME'] = 'Xen'

    if 'OBJECT_PREFIX' not in os.environ.keys():
        os.environ['OBJECT_PREFIX'] = 'XenProject'

    os.environ['MAJOR_VERSION'] = '8'
    os.environ['MINOR_VERSION'] = '1'
    os.environ['MICRO_VERSION'] = '0'

    if 'BUILD_NUMBER' not in os.environ.keys():
        os.environ['BUILD_NUMBER'] = next_build_number()

    print("BUILD_NUMBER=%s" % os.environ['BUILD_NUMBER'])

    if 'GIT_REVISION' in os.environ.keys():
        revision = open('revision', 'w')
        print(os.environ['GIT_REVISION'], file=revision)
        revision.close()

    #make_header()

    #copy_mof(driver)

    #copy_inf(driver, vs)

    symstore_del(driver, 30)

    if vs=='vs2012':
        release = 'Windows Vista'
    else:
        release = 'Windows 7'

    build_sln(driver, release, 'x86', debug[sys.argv[1]], vs)
    build_sln(driver, release, 'x64', debug[sys.argv[1]], vs)

    symstore_add(driver, release, 'x86', debug[sys.argv[1]], vs)
    symstore_add(driver, release, 'x64', debug[sys.argv[1]], vs)

    if len(sys.argv) <= 2 or sdv[sys.argv[2]]:
        run_sdv('xeniface', driver, vs)

    archive(driver + '\\source.tgz', manifest().splitlines(), tgz=True)
    archive(driver + '.tar', [driver,'revision'])

